### ISSUE TITLE
Treesitter formatexpr

### ISSUE THREAD

**Issue by kyazdani42** [nvim-treesitter/nvim-treesitter#613]
i found this really cool formatexpr option (which is local to buffer), which might be done using treesitter.
This might allow us not to rely on external formatters while writing the code in neovim.
not sure if we should do this here or in an external repo, because it might be a lot of code to add in.

**Comment by vigoux** [comment]
I think this can be the occasion of moving all the `foldexpr` ìndentexpr` to a separate module (`nvim-treesitter-builtins` maybe ?) and add more things there, this would allow people to pull it only if they want it.

**Comment by theHamsta** [comment]
What's the difference between indentexpr and formatexpr? How incomplete lines are handled? Doesn't formatexpr just do indentation? 

I thought of having a tree-sitter pretty printer. But this task is so much better handled by LSP that I quickly dropped the idea.

**Comment by kyazdani42** [comment]
indentexpr only indents the lines
formatexpr format the code. For instance if you had something like
```lua
function v()
call_some_func(); end
```
then formatexpr would remove the `;` and put the `end` keyword on the bottom \
indentexpr would only indent the 2nd line

**Comment by theHamsta** [comment]
So effectively a CST pretty printer. Any reason not to use LSP for this highly opionated/language-specific task (apart from Lua LSP not implementing this feature). A CST pretty printer would be useful if other guy plan to use tree-sitter for refactoring task or when you directly want to manipulate the CST and have the output printed in a nice manner. But probably still better to directly pipe result in lsp format_range.

**Comment by kyazdani42** [comment]
i like to think its better to use treesitter to do most of syntax related things (highlighting, indenting and formatting) and structural editing (like textobjects). To me, i tend to not like lsp a lot, because there are many issues with it, and the servers don't have the same features usually, which completely breaks my workflow... but it could accomplish things like `go_to_definition`, `diagnostics`, `references` and `completion` much better imo.

**Comment by theHamsta** [comment]
Usually every repo has some configuration which or sometimes even the whole language has this one tool that tells what is the correct formatting and no deviation is tolerable, which only LSP can do at the moment.

If you want to do the work, tree-sitter could be a as-you-type formatter (LSP might be to delayish for this) or a formatter that gives some ok result that does if LSP is not available. For me, even for Lua it would be much more convenient to just pipe the file into a language specific tool that to use a tree-sitter.

For implementation, one could just pretty print the CST: a space between every node, line breaks for nodes that are marked to require line breaks, indents from indentexpr (1). Before pretty-printing, one could a apply some beautification rules (like drop `;` when it separates statements) (2) and line break transformations for too long nodes (3). I think this could be useful if we had general facility to describe and apply AST transformations (2). (1,3) could be implemented without to much effort as a MVP. (2) would need some good description DSL to describe the transform rules. Maybe a external plugin could use fennel pattern matching for it https://fennel-lang.org/reference#match-pattern-matching or use queries with directives. In you example, one could write a query that matches your `;` and call `remove!` on it which would dispatch to a lua function that does some text processing.

**Comment by kyazdani42** [comment]
closing this since this is old and don't matter very much anyway

**Comment by clason** [comment]
For the sake of completeness: I do think tree-sitter is the right tool -- or rather level -- for this rather than LSP, as formatting is a syntactic operation that does not require project-wide (cross-file) "intelligence". The fact that many LSPs offer formatting is purely historical, coming from its root in language-specific VS Code extensions (and because they need to parse the syntax of the indexed files, anyway, be it using tree-sitter or some other tool).

That being said, the actual meat of the operation is specifying _how_ to format the code, and this is a highly language-specific and opinionated question. I do _not_ want to open that can of worms; we have more than enough on our plate already.

So TL;DR: I think a tree-sitter based formatting "meta-plugin" that allows (relatively) easily implementing formatters for languages with (sufficiently good) tree-sitter parsers would be very interesting, but that plugin should not be `nvim-treesitter` (at least until we have sorted out highlighting, folding, and indentation to the point where it's stable enough to be upstreamed).
